drivers/hmis/app/graphql/mutations/update_external_form_submission.rb

```diff
@@ -8,24 +8,72 @@ module Mutations
   class UpdateExternalFormSubmission < CleanBaseMutation
     argument :input, Types::HmisSchema::ExternalFormSubmissionInput, required: true
     argument :id, ID, required: true
+    argument :project_id, ID, required: false # not required for backwards compatibility
 
-    field :external_form_submission, Types::HmisSchema::ExternalFormSubmission, null: false
+    field :external_form_submission, Types::HmisSchema::ExternalFormSubmission, null: true
     field :errors, [Types::HmisSchema::ValidationError], null: false, resolver: Resolvers::ValidationErrors
 
-    def resolve(id:, input:)
+    def resolve(...)
+      Hmis::Hud::Base.transaction do
+        _resolve(...)
+      end
+    end
+
+    protected
+
+    EXTRANEOUS_KEYS = ['captcha_score', 'form_definition_id', 'form_content_digest'].freeze
+
+    def _resolve(id:, project_id:, input:)
       record = HmisExternalApis::ExternalForms::FormSubmission.find(id)
       raise 'Access denied' unless allowed?(permissions: [:can_manage_external_form_submissions])
 
       record.assign_attributes(**input.to_params)
 
-      errors = []
-      if record.valid?
-        record.save!
-      else
-        errors = record.errors
-        record = nil
+      errors = HmisErrors::Errors.new
+
+      unless record.valid?
+        errors.add_ar_errors(record.errors&.errors)
+        return { errors: errors }
       end
 
+      if record.status_changed? && record.status == 'reviewed' && !record.spam
+        definition = record.definition
+
+        # Only if there are Client and/or Enrollment fields in the form definition, initialize an enrollment
+        # (which will in turn initialize a Client, inside the form processor).
+        if definition.link_id_item_hash.values.find { |item| ['ENROLLMENT', 'CLIENT'].include?(item.mapping.record_type) }
+          project = Hmis::Hud::Project.find(project_id)
+          record.build_enrollment(project: project, data_source: project.data_source, entry_date: record.created_at)
+          # Assume that required values on Client and Enrollment (e.g. relationship to HoH) are present
+          # and correctly mapped in raw_values. The form processor record validator will fail otherwise.
+        end
+
+        form_processor = record.form_processor || record.build_form_processor(definition: definition)
```

Comment:
In what situation would the record already have a form processor? Are we supporting re-processing submissions? (That might be a good idea, actually, just didn't think it was part of the ux)

Reply:
Good point, I was copying from the submit assessment mutation.

Right now though, the UX doesn't prevent this:
- Review a form submission
- Open the reviewed submission again, uncheck "Mark reviewed" and hit "Save"
- Open the un-reviewed submission again, check "Mark reviewed" and hit "Save"

I'm seeing this leads to a deadlock error so I can investigate more
<img width="893" alt="Screenshot 2024-09-04 at 4 35 02 PM" src="https://github.com/user-attachments/assets/76b95599-0b45-4019-80da-ff285ee79618">

Reply:
Hm, the deadlock is not specific to the re-review case, it turns out. I suspect maybe introduced here: https://github.com/greenriver/hmis-warehouse/pull/4657#discussion_r1744714593, I'll keep investigating.

anyway, I also added test coverage for the re-reviewing situation. https://github.com/greenriver/hmis-warehouse/pull/4657/commits/429c464f06bd9542eee80eefa08d20504264cbac

Reply:
Ahmm, okay. I don't think we should allow "un-review".. we don't allow anything like that for form submission (going from submitted back to WIP), and I think the desired behavior is unclear (eg should the client be deleted?). I didn't realize the UI permitted that currently. Do you have a use-case in mind for keeping it?

If you're good with removing it, we can do the work to disable the checkbox in a separate issue, but here it probably makes sense to add a check in `UpdateExternalFormSubmission` to ensure that a reviewed submission can't get marked as unreviewed.

Reply:
This would be a quick solution to add on FormSubmission, which the mutation will handle returning:
```
validate :validate_status_change
private def validate_status_change
  return unless status_changed? && status.present? && status_was.present?

  errors.add(:status, :invalid, full_message: 'Cannot change status from Reviewed to New') if status_was == 'reviewed' && status == 'new'
end
```

Reply:
Ok perfect. I agree, there is no use case for it. It was just leftover from when 'reviewing' a form submission didn't really do anything aside from changing the attribute on the record.

Reply:
I added a bullet to disable the checkbox to this frontend ticket: https://github.com/open-path/Green-River/issues/6598